### ISSUE TITLE
PL_FAULT_TOLERANT_TRAINING=1 causes TypeError on next epoch after resuming

### ISSUE THREAD

**Issue by SpontaneousDuck** [Lightning-AI/lightning#13843]
## 🐛 Bug

After resuming training on a checkpoint saved using PL_FAULT_TOLERANT_TRAINING (specifically a resumed SLURM auto-checkpoint in my case), the training successfully resumes but then crashes with the below error on the start of the next epoch:

```
  File "/home/k.witham/source/rf_work/test/train.py", line 89, in <module>
    main()
  File "/home/k.witham/source/rf_work/test/train.py", line 80, in main
    train(model)
  File "/home/k.witham/source/rf_work/test/train.py", line 60, in train
    trainer.fit(model, datamodule=dm)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 770, in fit
    self._call_and_handle_interrupt(
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 723, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 811, in _fit_impl
    results = self._run(model, ckpt_path=self.ckpt_path)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1236, in _run
    results = self._run_stage()
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1323, in _run_stage
    return self._run_train()
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1353, in _run_train
    self.fit_loop.run()
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/loops/base.py", line 204, in run
    self.advance(*args, **kwargs)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/loops/fit_loop.py", line 266, in advance
    self._outputs = self.epoch_loop.run(self._data_fetcher)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/loops/base.py", line 204, in run
    self.advance(*args, **kwargs)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 171, in advance
    batch = next(data_fetcher)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/utilities/fetching.py", line 184, in __next__
    return self.fetching_function()
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/utilities/fetching.py", line 259, in fetching_function
    self._fetch_next_batch(self.dataloader_iter)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/utilities/fetching.py", line 273, in _fetch_next_batch
    batch = next(iterator)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/trainer/supporters.py", line 558, in __next__
    return self.request_next_batch(self.loader_iters)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/trainer/supporters.py", line 570, in request_next_batch
    return apply_to_collection(loader_iters, Iterator, next)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/utilities/apply_func.py", line 99, in apply_to_collection
    return function(data, *args, **kwargs)
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 530, in __next__
    data = self._next_data()
  File "/home/k.witham/miniconda3/envs/test/lib/python3.10/site-packages/pytorch_lightning/utilities/auto_restart.py", line 440, in wrapper
    batch, state = combined_batch["data"], combined_batch[AutoRestartBatchKeys.PL_RESTART_META]
TypeError: list indices must be integers or slices, not str
```

I am using a Map Dataset and am using `num_workers=1` in my DataLoaders. It appears that `combined_batch` is just a returned batch and has not been turned in to a Dict with the "data" and AutoRestartBatchKeys keys added yet. 

### To Reproduce

Use PL_FAULT_TOLERANT_TRAINING=1 on any training with a MapDataset and kill the training at some point. Resume from that checkpoint and wait till the next epoch starts.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): 1.6.5
- PyTorch Version (e.g., 1.10): 1.11
- Python version (e.g., 3.9): 3.10.5
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: 11.3
- GPU models and configuration: Tesla P100 1
- How you installed PyTorch (`conda`, `pip`, source): conda

cc @carmocca @justusschock @awaelchli @ninginthecloud @akihironitta

**Comment by jhoareau** [comment]
The same problem is happening in 1.7.4.

**Comment by jhoareau** [comment]
The same problem is happening in 1.7.4, without using a SLURM cluster.

The issue seems to be because, in `_next_data_wrapper` in `auto_restart.py`:
- during epoch 0`fn` is of the type `<bound method _MultiProcessingDataLoaderIter._next_data of <torch.utils.data.dataloader._MultiProcessingDataLoaderIter object>>`
- in epoch 1, `fn` is of the type `<function _MultiProcessingDataLoaderIter._next_data>`

I dug a bit more and tracked where the type change seems to happen: line 178 of `utilities/fetching.py`
- during epoch 0, as `iter` is called once, `self.dataloader_iter.loader_iters._next_data` is the correct bound method
- during epoch 1, `iter` is called again and `self.dataloader_iter.loader_iters._next_data` becomes the wrong non-bound method

The behaviour in epoch 1 is quite weird, I don't know why the method becomes suddenly unbound while iterating on the `dataloader` object, which keeps the right type all along.

@tchaton @carmocca @awaelchli

**Comment by carmocca** [comment]
Thanks for the exploration @jhoareau! This is very valuable.

We will be looking at these (complex) fault-tolerance issues soon but we are prioritizing other features at the moment. If we cannot quickly solve these within this week, it is likely that they will need to wait for a month or less.

Sorry for the inconvenience.

**Comment by jhoareau** [comment]
This can be closed, fault tolerant support was dropped from PL 2.0

**Comment by awaelchli** [comment]
@jhoareau In principle yes. We haven't figured out how to address the problems mentioned here, and it can't be fixed in 2.0 because as you say, we removed the code. 

However note, we do have the commitment to support 1.9.x long-term, and so bug fixes can also happen for that version independently. If we find that more users depend on this fix, we could bump the priority. But I think the chances are low that this can be completed. Let me know what you think.
